File: A2A-Star-Algorithm/AStarEasy.java
import java.util.*;

class AStarEasy 
{

    static class Node implements Comparable<Node> 
    {
        String state;
        int g, h;
        Node parent;

        Node(String s, int g, int h, Node p) 
        {
            state = s;
            this.g = g;
            this.h = h;
            parent = p;
        }

        int f() 
        {
            return g + h;
        }

        public int compareTo(Node n) 
        {
            return this.f() - n.f();
        }
    }

    static Map<String, List<String>> graph = new HashMap<>();
    static Map<String, Integer> cost = new HashMap<>();

    static void addEdge(String a, String b, int c) 
    {
        graph.putIfAbsent(a, new ArrayList<>());
        graph.get(a).add(b);
        cost.put(a + b, c);
    }

    static void aStar(String start, String goal,
                      Map<String, Integer> hVal) 
    {
        System.out.println("Start Node: " + start);
        System.out.println("Goal Node : " + goal);
        System.out.println();

        PriorityQueue<Node> open = new PriorityQueue<>();
        Set<String> closed = new HashSet<>();

        open.add(new Node(start, 0, hVal.get(start), null));

        while (!open.isEmpty()) 
        {

            Node cur = open.poll();

            if (cur.state.equals(goal)) 
            {
                System.out.print("Path Found: ");
                print(cur);
                System.out.println("\nTotal Cost: " + cur.g);
                return;
            }

            closed.add(cur.state);

            for (String next : graph.getOrDefault(cur.state, new ArrayList<>())) 
            {
                if (!closed.contains(next)) 
                {
                    int g = cur.g + cost.get(cur.state + next);
                    int h = hVal.get(next);
                    open.add(new Node(next, g, h, cur));
                }
            }
        }

        System.out.println("No Path Found");
    }

    static void print(Node n) 
    {
        if (n == null) return;
        print(n.parent);
        System.out.print(n.state + " ");
    }

    public static void main(String[] args) 
    {

        addEdge("A", "B", 1);
        addEdge("A", "C", 2);
        addEdge("B", "D", 3);
        addEdge("C", "E", 5);
        addEdge("D", "F", 2);
        addEdge("F", "G", 1);
        addEdge("E", "G", 3);

        Map<String, Integer> h = Map.of(
            "A", 6,
            "B", 5,
            "C", 4,
            "D", 3,
            "E", 3,
            "F", 1,
            "G", 0
        );

        aStar("A", "G", h);
    }
}

File: A2A-Star-Algorithm/AstarEasy.py
import heapq


class Node:

    def __init__(self, state, g, h, parent):

        self.state = state
        self.g = g
        self.h = h
        self.parent = parent

    def f(self):
        return self.g + self.h

    def __lt__(self, other):
        return self.f() < other.f()


graph = {}
cost = {}


def add_edge(a, b, c):

    if a not in graph:
        graph[a] = []

    graph[a].append(b)

    cost[a + b] = c


def print_path(node):

    if node is None:
        return

    print_path(node.parent)

    print(node.state, end=" ")


def a_star(start, goal, h_val):

    print("Start Node:", start)
    print("Goal Node :", goal)
    print()

    open_list = []

    closed = set()

    heapq.heappush(open_list,
                   Node(start, 0, h_val[start], None))

    while open_list:

        cur = heapq.heappop(open_list)

        if cur.state == goal:

            print("Path Found:", end=" ")

            print_path(cur)

            print("\nTotal Cost:", cur.g)

            return

        closed.add(cur.state)

        for next_node in graph.get(cur.state, []):

            if next_node not in closed:

                g = cur.g + cost[cur.state + next_node]

                h = h_val[next_node]

                heapq.heappush(
                    open_list,
                    Node(next_node, g, h, cur)
                )

    print("No Path Found")


# Main Program

add_edge("A", "B", 1)
add_edge("A", "C", 2)
add_edge("B", "D", 3)
add_edge("C", "E", 5)
add_edge("D", "F", 2)
add_edge("F", "G", 1)
add_edge("E", "G", 3)

h = {
    "A": 6,
    "B": 5,
    "C": 4,
    "D": 3,
    "E": 3,
    "F": 1,
    "G": 0
}

a_star("A", "G", h)